### ISSUE TITLE
The lastest torch.CudaTensor.multinomial() sometimes return 0 

### ISSUE THREAD

**Issue by doubledaibo** [torch/cutorch#700]
The second assertion in the following code will fail occasionally, while the first assertion succeed.
   assert(torch.sum(out) > 0)
   it = torch.multinomial(out, 1):view(-1)
   assert(torch.sum(torch.eq(it, 0)) == 0)

Save the snapshot and minibatch right before the error cannot reproduce the error. 
But run from the start will cause the error at the same iteration. (same random seed)
Trying to get a minimum fail case.

**Comment by doubledaibo** [comment]
Changing random seed sometimes cause another error:

/tmp/luarocks_cutorch-scm-1-5865/cutorch/lib/THC/THCTensorIndex.cu:321: void indexSelectLargeIndex(TensorInfo<T, IndexType>, TensorInfo<T, IndexType>, TensorInfo<long, In
dexType>, int, int, IndexType, IndexType, long) [with T = float, IndexType = unsigned int, DstDim = 2, SrcDim = 2, IdxDim = -2]: block: [74,0,0], thread: [32,0,0] Asserti
on `srcIndex < srcSelectDimSize` failed.
/tmp/luarocks_cutorch-scm-1-5865/cutorch/lib/THC/THCTensorIndex.cu:321: void indexSelectLargeIndex(TensorInfo<T, IndexType>, TensorInfo<T, IndexType>, TensorInfo<long, In
dexType>, int, int, IndexType, IndexType, long) [with T = float, IndexType = unsigned int, DstDim = 2, SrcDim = 2, IdxDim = -2]: block: [74,0,0], thread: [33,0,0] Asserti
on `srcIndex < srcSelectDimSize` failed.
/tmp/luarocks_cutorch-scm-1-5865/cutorch/lib/THC/THCTensorIndex.cu:321: void indexSelectLargeIndex(TensorInfo<T, IndexType>, TensorInfo<T, IndexType>, TensorInfo<long, In
dexType>, int, int, IndexType, IndexType, long) [with T = float, IndexType = unsigned int, DstDim = 2, SrcDim = 2, IdxDim = -2]: block: [74,0,0], thread: [34,0,0] Asserti
on `srcIndex < srcSelectDimSize` failed.
... (repeat for blocks [70-75,0,0], thread: [32-127,0,0])

THCudaCheck FAIL file=/tmp/luarocks_cutorch-scm-1-5865/cutorch/lib/THC/generated/../generic/THCTensorMathReduce.cu line=262 error=59 : device-side assert triggered
/data2/bdai/torch/install/bin/luajit: cuda runtime error (59) : device-side assert triggered at /tmp/luarocks_cutorch-scm-1-5865/cutorch/lib/THC/generated/../generic/THCT
ensorMathReduce.cu:262

stack traceback:
        [C]: at 0x7fba5abab560
        [C]: in function 'sum'
the line is 
  
     if torch.sum(it) == 0 then

**Comment by killeent** [comment]
@doubledaibo can you provide a script that reproduces this error consistently?

**Comment by pavanky** [comment]
This is not new: https://github.com/torch/cutorch/issues/398

**Comment by doubledaibo** [comment]
@pavanky Related but slightly different, #398 says multinomial will sample some item which has zero probability, here I found multinomial sometimes sample 0 instead of some value in [1..n].

**Comment by doubledaibo** [comment]
@killeent I'm trying to find a minimum fail case.
